File: internal/lifecycle/rename.go
package lifecycle

import (
	"context"
	"errors"
	"fmt"

	"github.com/abbyfluoroethane/bento/internal/hypervisor"
	"github.com/abbyfluoroethane/bento/internal/types"
)

// ErrRenameNeedsStop rejects a rename of a running instance: the libvirt
// domain carries the instance name (SPEC 7), and a defined domain can
// only change its name through an undefine and redefine, which needs the
// domain off.
var ErrRenameNeedsStop = errors.New("lifecycle: stop the instance before renaming it")

// Rename changes the name of an instance. The old name enters the
// release cooldown; no alias and no redirect is created (SPEC 7.2, 7.3).
// The overlay and seed ISO paths derive from the UUID and never move;
// the libvirt domain is undefined under the old name and redefined under
// the new one, so the instance must be stopped.
func (m *Manager) Rename(ctx context.Context, uuid, newName string) error {
	inst, err := m.store.Instance(uuid)
	if err != nil {
		return err
	}
	if newName == inst.Name {
		return nil
	}
	definer, canDefine := m.hyp.(Definer)

	// Ask libvirt, not the possibly stale state column: destroying a
	// running domain here would be a surprise stop.
	oldName := inst.Name
	domainGone := false
	switch state, err := m.hyp.State(ctx, oldName); {
	case errors.Is(err, hypervisor.ErrDomainNotFound):
		domainGone = true // row without domain; reconcile reports it
	case err != nil:
		return fmt.Errorf("lifecycle: rename %s: %w", oldName, err)
	case state != types.StateStopped:
		return fmt.Errorf("%w: %s is %s", ErrRenameNeedsStop, oldName, state)
	}
	if !domainGone && !canDefine {
		return fmt.Errorf("lifecycle: rename %s: hypervisor cannot redefine domains", oldName)
	}

	// The store enforces the unique name and the cooldown of the new
	// name in one transaction (SPEC 7.2).
	if err := m.store.RenameInstance(uuid, newName, m.cooldown); err != nil {
		return err
	}
	if domainGone {
		return nil
	}

	inst.Name = newName
	withISO := m.isoExists(m.SeedISOPath(uuid))
	xml, err := m.domainXMLByUUID(inst, withISO)
	if err != nil {
		return m.unwindRename(inst, oldName, err)
	}
	if err := m.hyp.Remove(ctx, oldName); err != nil && !errors.Is(err, hypervisor.ErrDomainNotFound) {
		return m.unwindRename(inst, oldName, fmt.Errorf("undefine %s: %w", oldName, err))
	}
	if err := definer.Define(ctx, xml); err != nil {
		// The old definition is gone. Try to bring it back under the
		// old name before reverting the row.
		inst.Name = oldName
		if oldXML, xmlErr := m.domainXMLByUUID(inst, withISO); xmlErr == nil {
			if defErr := definer.Define(ctx, oldXML); defErr != nil {
				m.log.Error("rename: domain lost; redefine it by hand",
					"instance", oldName, "uuid", uuid, "error", defErr)
			}
		}
		inst.Name = newName
		return m.unwindRename(inst, oldName, fmt.Errorf("define %s: %w", newName, err))
	}
	m.log.Info("rename: instance renamed", "from", oldName, "to", newName, "uuid", uuid)
	return nil
}

// unwindRename reverts the row to the old name after a failed domain
// redefine. The owner retakes a name from their own cooldown at once
// (SPEC 7.2 rule 1), so the revert cannot be blocked.
func (m *Manager) unwindRename(inst types.Instance, oldName string, cause error) error {
	if err := m.store.RenameInstance(inst.UUID, oldName, m.cooldown); err != nil {
		m.log.Error("rename: revert to the old name failed",
			"instance", inst.Name, "old", oldName, "error", err)
	}
	return fmt.Errorf("lifecycle: rename %s to %s: %w", oldName, inst.Name, cause)
}

File: crates/lifecycle/src/rename.rs
use bento_hypervisor::Error as HypervisorError;
use bento_types::State;

use crate::{Error, Manager, Result};

impl Manager {
    /// Changes an instance label without moving its UUID-derived disk or seed.
    /// The stopped domain is undefined under the old name and redefined under
    /// the new one; the old name enters cooldown with no alias (SPEC 7.2, 7.3).
    pub async fn rename(&self, uuid: &str, new_name: &str) -> Result<()> {
        let mut instance = self
            .store
            .instance(uuid)
            .await
            .map_err(crate::actions::external)?;
        if new_name == instance.name {
            return Ok(());
        }
        let old_name = instance.name.clone();
        let domain_gone = match self.hyp.state(&old_name).await {
            Ok(State::Stopped) => false,
            Ok(state) => return Err(Error::RenameNeedsStop(format!("{old_name} is {state}"))),
            Err(HypervisorError::DomainNotFound(_)) => true,
            Err(error) => {
                return Err(Error::operation(format!(
                    "lifecycle: rename {old_name}: {error}"
                )));
            }
        };
        if !domain_gone && self.definer.is_none() {
            return Err(Error::operation(format!(
                "lifecycle: rename {old_name}: hypervisor cannot redefine domains"
            )));
        }
        self.store
            .rename_instance(uuid, new_name, self.cooldown)
            .await
            .map_err(crate::actions::external)?;
        if domain_gone {
            return Ok(());
        }
        instance.name = new_name.to_string();
        let with_iso = (self.iso_exists)(&self.seed_iso_path(uuid));
        let xml = match self.domain_xml_by_uuid(&instance, with_iso).await {
            Ok(xml) => xml,
            Err(error) => return Err(self.unwind_rename(&instance, &old_name, error).await),
        };
        match self.hyp.remove(&old_name).await {
            Ok(()) | Err(HypervisorError::DomainNotFound(_)) => {}
            Err(error) => {
                return Err(self
                    .unwind_rename(
                        &instance,
                        &old_name,
                        Error::operation(format!("undefine {old_name}: {error}")),
                    )
                    .await);
            }
        }
        let definer = self.definer.as_ref().expect("capability checked");
        if let Err(error) = definer.define(&xml).await {
            let mut old = instance.clone();
            old.name.clone_from(&old_name);
            if let Ok(old_xml) = self.domain_xml_by_uuid(&old, with_iso).await
                && let Err(restore_error) = definer.define(&old_xml).await
            {
                self.log.error(&format!(
                    "rename: domain lost; redefine it by hand: {restore_error}"
                ));
            }
            return Err(self
                .unwind_rename(
                    &instance,
                    &old_name,
                    Error::operation(format!("define {new_name}: {error}")),
                )
                .await);
        }
        Ok(())
    }

    async fn unwind_rename(
        &self,
        instance: &bento_types::Instance,
        old_name: &str,
        cause: Error,
    ) -> Error {
        if let Err(error) = self
            .store
            .rename_instance(&instance.uuid, old_name, self.cooldown)
            .await
        {
            self.log
                .error(&format!("rename: revert to old name failed: {error}"));
        }
        Error::operation(format!(
            "lifecycle: rename {old_name} to {}: {cause}",
            instance.name
        ))
    }
}
